iceoryx2-bb/container/Cargo.toml

```diff
@@ -14,7 +14,7 @@ version = { workspace = true }
 
 [dependencies]
 iceoryx2-bb-derive-macros = { workspace = true }
-iceoryx2-bb-elementary = { workspace = true }
+iceoryx2-bb-elementary = { workspace = true , features = ["iox2-test"]}
```

Comment:
This will make the features available to all user of `iceoryx2-bb-elementary`, at least in this workspace. I'm not sure what happens with external users. See also https://doc.rust-lang.org/cargo/reference/features.html#feature-unification

Could you check if an external crate that depends on `container` and `elementary` has access to the `BumpAllocator` without specifying the feature flag itself?

A workaround could be to add `iceoryx2-bb-elementary = { workspace = true , features = ["iox2-test"]}` as dev dependency and leave the regular dependency without the flag. One could even go one step further and just add `iceoryx2-bb-elementary` with the flag as regular dependency to `iox2-bb-testing` and re-export it in that module. Since there would be no transitive dependency on a `iceoryx-bb-elementary` with the flag set, except the dev targets, regular users of the crate should not have access to the `BumpAllocator` by accident.

What do you think?

I still need to test this with bazel tomorrow.

Reply:
My bad, the workspace usage actually enables this feature to all users. But we can enable the feature inside the `dev-dependency` so it could solve this issue here perfectly.

If an external user runs `cargo build`, rustc fails and reports the errors. This satisfy our requirement:
- internal testing cases doesn't be affected
- outside users cannot use it except explicit enable the `testing` feature

```
error[E0433]: failed to resolve: could not find `bump_allocator` in `iceoryx2_bb_elementary`
  --> src/main.rs:6:36
   |
6  |     let _a=iceoryx2_bb_elementary::bump_allocator::BumpAllocator::new(8);
   |                                    ^^^^^^^^^^^^^^ could not find `bump_allocator` in `iceoryx2_bb_elementary`
   |
note: found an item that was configured out
  --> /Users/yuchen.xie/.cargo/git/checkouts/iceoryx2-d2f97997d049ef12/e3225b5/iceoryx2-bb/elementary/src/lib.rs:23:9
   |
23 | pub mod bump_allocator;
   |         ^^^^^^^^^^^^^^
note: the item is gated behind the `testing` feature
  --> /Users/yuchen.xie/.cargo/git/checkouts/iceoryx2-d2f97997d049ef12/e3225b5/iceoryx2-bb/elementary/src/lib.rs:22:7
   |
22 | #[cfg(feature = "testing")]
```

### external user cargo toml 

```
[package]
name = "iceoryx2-playground"
version = "0.1.0"
edition = "2021"

[dependencies]
iceoryx2 = { git = "https://github.com/xieyuschen/iceoryx2", rev = "e3225b55" }
iceoryx2-bb-elementary = { git = "https://github.com/xieyuschen/iceoryx2", rev = "e3225b55" }
```
### main.rs

```rust
use core::time::Duration;
use iceoryx2::prelude::*;
const CYCLE_TIME: Duration = Duration::from_secs(1);

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let _a=iceoryx2_bb_elementary::bump_allocator::BumpAllocator::new(8);
    let node = NodeBuilder::new().create::<ipc::Service>()?;

    let service = node.service_builder(&"My/Funk/ServiceName".try_into()?)
        .publish_subscribe::<usize>()
        .open_or_create()?;

    let publisher = service.publisher_builder().create()?;

    while node.wait(CYCLE_TIME).is_ok() {
        let sample = publisher.loan_uninit()?;
        let sample = sample.write_payload(1234);
        sample.send()?;
    }

    Ok(())
}
```